src/utils/types.ts

```diff
@@ -22,3 +22,33 @@ export type RequiredId<T extends { [key in "id"]?: string }> = Replace<T, "id",
 export type Replace<T, Key extends keyof T, NewValue, IsOptional extends boolean = false> =
   & Omit<T, Key>
   & (IsOptional extends true ? { readonly [key in Key]?: NewValue } : { readonly [key in Key]: NewValue });
+
+/**
+ * Extracts a transformed `ReadonlyMap` where the values are simplified to the `selfId` property
+ * of the objects if available. Otherwise returns the type as is.
+ *
+ * @template T - The type to transform, e.g. `ReadonlyMap<string, object>`.
+ * @returns A `ReadonlyMap` keyed similarly but with values as the `selfId` of the objects,
+ *          or the original type `T` if `T` is not a `ReadonlyMap` or the objects don't have `selfId`.
+ */
+export type ExtractSelfId<T> = T extends ReadonlyMap<string, infer U>
+  ? U extends { selfId: infer V } ? ReadonlyMap<string, V>
```

Comment:
self id is always string :D I just have feeling this is kind of overengineering

Reply:
don't laugh at me :( but yeah, the inference is redundant... I did that on purpose, to see if you review thoroughly... :D